### PR TITLE
Added parameter range checks for all optimizers

### PR DESCRIPTION
**Pull request by lazypanda1**
This PR adds parameter range checks to all optimizers to ensure that end-users do not end up providing invalid values to the optimizers and be confused by the output when there is no actual problem with their model.

For example, running the following program produces `NaN`s in the output, due to invalid value of `rho` (>1.0).

```python
import torch
from torch.autograd import Variable

N, D_in, H, D_out = 64, 1000, 100, 10

x = Variable(torch.randn(N, D_in))
y = Variable(torch.randn(N, D_out), requires_grad=False)

model = torch.nn.Sequential(
    torch.nn.Linear(D_in, H),
    torch.nn.ReLU(),
    torch.nn.Linear(H, D_out),
)
loss_fn = torch.nn.MSELoss(size_average=False)

learning_rate = 1e-4
optimizer = torch.optim.Adadelta(model.parameters(), lr=learning_rate, rho=1.1)
for t in range(2):
    y_pred = model(x)
    loss = loss_fn(y_pred, y)
    print(t, loss.data[0])
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
```

Output:
```
0 651.8707885742188
1 nan
```

I tried adding constraints for all the parameters that I could infer from the corresponding articles, but I am still missing some. Please feel free to suggest what should be bound for the ones which are missing.

This is similar to the bounds check which I added for [`Adam Optimizer`](https://github.com/pytorch/pytorch/pull/5147)

I can also add tests if needed.

### CODE REVIEW

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -50,6 +50,13 @@ class SGD(Optimizer):
 
     def __init__(self, params, lr=required, momentum=0, dampening=0,
                  weight_decay=0, nesterov=False):
+        if not 0.0 <= lr:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if not 0.0 <= momentum <= 1.0:
+            raise ValueError("Invalid momentum value: {}".format(momentum))
```
Comment (vishwakftw) [inline torch/optim/sgd.py]: This is questionable. In practice only a momentum value less than 1 is used, but should not restrict the user from having higher momentum.

Comment (lazypanda1) [reply torch/optim/sgd.py]: Agreed. Will change this. Can you also review other ranges?

**Review by vishwakftw — APPROVED** [review]
I have reviewed the rest of the ranges. Looks good to me.

**Review by apaszke — APPROVED** [review]

File: test/test_optim.py [hunk 2]
```diff
@@ -343,6 +349,8 @@ def test_adagrad(self):
                 self._build_params_dict(weight, bias, lr=1e-2),
                 lr=1e-1)
         )
+        with self.assertRaisesRegex(ValueError, "Invalid lr_decay value: -0.5"):
+            optim.Adagrad(None, lr=1e-2, lr_decay=-0.5)
 
     def test_adagrad_sparse(self):
         self._test_rosenbrock_sparse(
```
Comment (lazypanda1) [inline test/test_optim.py]: for `test_adamax`below, some test cases use `optim.Adagrad` instead of `optim.Adamax`. Is this intentional? Or a bug?

Comment (ezyang) [reply test/test_optim.py]: Sounds like a bug to me.

Comment (lazypanda1) [reply test/test_optim.py]: I will address this in a separate PR

**Comment by lazypanda1** [comment]
Can this be merged?

**Review by apaszke — APPROVED** [review]

**Pull request merged**